web/src/message_store.ts

```diff
@@ -394,16 +400,204 @@ export function reify_message_id({old_id, new_id}: {old_id: number; new_id: numb
         message.locally_echoed = false;
         stored_messages.set(new_id, {type: "server_message", message});
         stored_messages.delete(old_id);
+        if (message.type === "stream") {
+            remove_topic_links(message.stream_id, message.topic, message.id);
+            save_topic_links(message);
+        }
+    }
+}
+
+export type TopicLink = {
+    // Where the link occurs
+    from_message_id: number;
+    from_stream_id: number;
+    from_topic: string;
+    from_link: string;
+    // What's being linked to
+    to_stream_id: number;
+    to_topic: string | undefined;
+    to_message_id: number | undefined;
+    to_link: string;
+};
+
+const topic_links_by_from = new Map<number, Map<string, Set<TopicLink>>>();
+const topic_links_by_to = new Map<number, Map<string, Set<TopicLink>>>();
```

Comment:
I went back and forth a lot on how to set up this data structure. This feels okay to me but still more involved than ideal. Let me know if you have better ideas!

Reply:
https://chat.zulip.org/#narrow/channel/6-frontend/topic/data.20structure.20for.20topic.20links/with/2284561